validator/client/validator.go

```diff
@@ -128,7 +134,105 @@ type attSelectionKey struct {
 
 // Done cleans up the validator.
 func (v *validator) Done() {
-	v.ticker.Done()
+	if v.accountChangedSub != nil {
+		v.accountChangedSub.Unsubscribe()
+	}
+	if v.ticker != nil {
+		v.ticker.Done()
+	}
+}
+
+func (v *validator) EventsChan() <-chan *eventClient.Event {
+	return v.eventsChannel
+}
+
+func (v *validator) AccountsChangedChan() <-chan [][fieldparams.BLSPubkeyLength]byte {
+	return v.accountsChangedChannel
+}
+
+func (v *validator) GenesisTime() uint64 {
+	return v.genesisTime
+}
+
+func (v *validator) Init(ctx context.Context) error {
+	ctx, span := trace.StartSpan(ctx, "validator.Init")
+	defer span.End()
+
+	ticker := time.NewTicker(backOffPeriod)
+	defer ticker.Stop()
+
+	firstTime := true
+	for {
+		if !firstTime {
+			if ctx.Err() != nil {
+				log.Info("Context canceled, stopping validator")
+				return errors.New("context canceled")
+			}
+			<-ticker.C
+		}
+
+		firstTime = false
+		if err := v.WaitForChainStart(ctx); err != nil {
+			if isConnectionError(err) {
+				log.WithError(err).Warn("Could not determine if beacon chain started")
+				continue
+			}
+
+			return errors.Wrap(err, "Could not determine if beacon chain started")
+		}
+
+		if err := v.WaitForKeymanagerInitialization(ctx); err != nil {
+			return errors.Wrap(err, "Wallet is not ready")
+		}
+
+		if err := v.WaitForSync(ctx); err != nil {
+			if isConnectionError(err) {
+				log.WithError(err).Warn("Could not determine if beacon chain started")
+				continue
+			}
+
+			return errors.Wrap(err, "Could not determine if beacon node synced")
+		}
+
+		if err := v.WaitForActivation(ctx); err != nil {
+			return errors.Wrap(err, "Could not wait for validator activation")
+		}
+
+		if err := v.CheckDoppelGanger(ctx); err != nil {
+			if isConnectionError(err) {
+				log.WithError(err).Warn("Could not wait for checking doppelganger")
+				continue
+			}
+
+			return errors.Wrap(err, "Could not succeed with doppelganger check")
+		}
+		break
+	}
+	currentSlot := slots.CurrentSlot(v.genesisTime) // set in v.WaitForChainStart
+	epochStart, err := slots.EpochStart(slots.ToEpoch(currentSlot) + 1)
+	if err != nil {
+		return errors.Wrapf(err, "Could not get epoch start from current slot %d", currentSlot)
+	}
+	if err := v.UpdateDuties(ctx); err != nil {
+		handleAssignmentError(err, epochStart)
+		return errors.Wrap(err, "Could not update duties")
```

Comment:
I changed it to require duties when you start up, currently you can enter the main loop without duties which leads to cosmetic errors, the downside to this is  you need to manually restart the validator if this doesn't complete because it should crash here. 

I'm not sure of the best option.